src/mongo_logger.ts

```diff
@@ -0,0 +1,144 @@
+
+/** @internal */
+export type MongoLoggableComponent =
+  typeof MongoLoggableComponent[keyof typeof MongoLoggableComponent];
+
+/** @internal */
+export interface MongoLoggerMongoClientOptions {
+  mongodbLogPath?: string | Writable;
+}
+
+/** @internal */
+export interface MongoLoggerOptions {
+  /** Severity level for command component */
+  command: SeverityLevel;
+  /** Severity level for SDAM */
+  topology: SeverityLevel;
+  /** Severity level for server selection component */
+  serverSelection: SeverityLevel;
+  /** Severity level for CMAP */
+  connection: SeverityLevel;
+  /** Default severity level to be if any of the above are unset */
+  defaultSeverity: SeverityLevel;
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  maxDocumentLength: number;
+  /** Destination for log messages. Must be 'stderr', 'stdout', a file path, or a Writable. Defaults to 'stderr'. */
+  logPath: string | Writable;
+}
+
+/** @internal */
+export class MongoLogger {
+  componentSeverities: Record<MongoLoggableComponent, SeverityLevel>;
+  maxDocumentLength: number;
+  logPath: Writable;
+
+  constructor(options: MongoLoggerOptions) {
+    // validate log path
+    if (typeof options.logPath === 'string') {
+      this.logPath =
+        options.logPath === 'stderr' || options.logPath === 'stdout'
+          ? process[options.logPath]
+          : fs.createWriteStream(options.logPath, { flags: 'a+' });
+      // TODO(NODE-4816): add error handling for creating a write stream
+    } else {
+      this.logPath = options.logPath;
+    }
+
+    this.componentSeverities = options;
+    this.maxDocumentLength = options.maxDocumentLength;
+  }
+
+  /* eslint-disable @typescript-eslint/no-unused-vars */
+  /* eslint-disable @typescript-eslint/no-empty-function */
+  emergency(component: any, message: any): void {}
+
+  alert(component: any, message: any): void {}
+
+  critical(component: any, message: any): void {}
+
+  error(component: any, message: any): void {}
+
+  warn(component: any, message: any): void {}
+
+  notice(component: any, message: any): void {}
+
+  info(component: any, message: any): void {}
+
+  debug(component: any, message: any): void {}
+
+  trace(component: any, message: any): void {}
+
+  /**
+   * Merges options set through environment variables and the MongoClient, preferring envariables
+   * when both are set, and substituting defaults for values not set. Options set in constructor
+   * take precedence over both environment variables and MongoClient options.
+   *
+   * @remarks
+   * When parsing component severity levels, invalid values are treated as unset and replaced with
+   * the default severity.
+   *
+   * @param clientOptions - options set for the logger in the MongoClient options
+   * @returns a MongoLoggerOptions object to be used when instantiating a new MongoLogger
+   */
+  static resolveOptions(clientOptions?: MongoLoggerMongoClientOptions): MongoLoggerOptions {
```

Comment:
neal and i spoke yesterday about moving away from using the global process and i think this would be a step in the right direction of making things more configurable and hopefully future-proofing this code

Reply:
Is the hesitation about using Document that it doesn't assert on which keys are on the object (like `MongoLoggerEnvOptions` does) or that it's incorrect?  You're right that Document isn't exactly correct because it allows arbitrary types for values (i.e., `{ 'name': 1234 }` is allowed but values can only be strings), either `Record<string, string | undefined>` or `NodeJS.ProcessEnv` would be the correct types.

I'll leave the choice up to you because I think both are fine.  But one benefit of passing in the `process.env` object directly and using either `Record<string, string | undefined>` or `NodeJS.ProcessEnv` as the type for the environment is that the `resolveOptions` function would encapsulate all options resolution.  The current approach spreads the options resolution across the `resolveOptions` function and parseOptions.